### DISCUSSION TITLE
[RFC] 卸载后重新渲染 Monaco 编辑器优化

### DISCUSSION THREAD

**Discussion by AhkunTa** [opensumi/core#2510] [RFC]
## 背景
OpenSumi 内使用 Monaco 编辑器作为基础编辑器，但 Monaco 编辑器为全局单例**并且只能这样**引入，初始化时 OpenSumi 会 override Monaco 内部的各个功能点如: 
- MonacoBulkEditService
- MonacoCodeService
- MonacoCommandService
- ScopedContextKeyService
- MonacoContextViewService
- OpenerService
- MonacoTextModelService

<img width="902" alt="截屏2023-03-28 14 48 18" src="https://user-images.githubusercontent.com/31676999/228151819-a0fdcf34-0696-46a3-9769-aaeb5198b5f7.png">

当作为组件嵌入页面时会存在组件销毁重新渲染，当重新渲染组件时 Monaco 内部并不会更新新的 override 的注入， 只会初始化一次[相关代码](https://github.com/opensumi/monaco-editor-core/blob/feat/monaco-core-0.35.x/src/vs/editor/standalone/browser/standaloneServices.ts#L1023-L1026)

这就会导致重新渲染后，再次编辑内部编辑器表现出各种问题 快捷键，交互，命令等。

### 1.修改 Monaco
不推荐改动，整体 Monaco 内部交互复杂，并且 Monaco 每次升级会伴随新的 break changes，会导致升级新版 Monaco 越来越困难

### 2.涉及的模块相应改造
打算将 Monaco 初始化 override 的模块改造为全局单例并且在销毁后 proxy 到最新的 injector 动态实例上，从而使用新的动态实例

## 改造清单
- 对应 override 的各个模块
- ...

## 测试范围
待补充

**Comment by erha19** [comment]
@AhkunTa 跟 Monaco 注册相关的问题还有这里的

- https://github.com/opensumi/core/discussions/2409

由于我们框架内的 Monaco 是在渲染阶段去引入并初始化的，会导致渲染时机过早而无法正确完成首次语言配置的识别，最终导致需要用户手动  `setLocale` 或首次语言配置不生效的问题。

因为这层逻辑，我们还在 `opensumi/monaco-editor-core` 内追加了读取 localStorage 语言配置的逻辑，这也是不合理的，可以一起看一下，这里是不是可以将 monaco 的加载/卸载逻辑一起做控制？

**Comment by AhkunTa** [comment]
### 主要问题

OpenSumi 内部注册 [OpenSumiContextKeyService](https://github.com/opensumi/core/blob/9a1e3685074dc0343b8e43470e290dfb6f52cb39/packages/monaco/src/browser/monaco.context-key.service.ts#L384) 来 override monaco 内部的 [MonacoContextKeyService](https://github.com/opensumi/monaco-editor-core/blob/a14c678ec28ec804b37984e07a794b9c22acf737/src/vs/platform/contextkey/browser/contextKeyService.ts#LL372C1-L372C1)

https://github.com/opensumi/core/blob/9a1e3685074dc0343b8e43470e290dfb6f52cb39/packages/editor/src/browser/editor.contribution.ts#L191-L215

而在 dispose 后重新渲染组件, `OpenSumiContextKeyService` 会在 OpenSumi 内重新实例化生成（内部的 contextId 从0开始生成）

但在 monaco 内部的 `MonacoContextKeyService`  在 dispose 会继续保存 （monaco只会实例化一次）

此时在opensumi 和 monaco 存在两个 `ContextKeyService`, 而 monaco 内部的一些 keybinding ,都会在 monaco 内寻找其对应的contextId ,通过在元素上寻找字符串 `data-keybinding-context`  

<img width="661" alt="截屏2023-05-18 20 21 42" src="https://github.com/opensumi/core/assets/31676999/aeff6366-3281-48d4-bbe3-99ec8ca5e16d">

如下图 寻找不到 37 的 contextId

<img width="502" alt="截屏2023-05-18 20 27 01" src="https://github.com/opensumi/core/assets/31676999/37374018-df10-44c1-8c70-c276b67b4f87">


### 解决方法
#### 1.全局单例 ContextKeyService 并在dispose 后获取 monaco 内部ContextKeyService 作为OpenSumi内新实例
#### 2.可能需要 注册 [StandaloneKeybindingService](https://github.com/opensumi/monaco-editor-core/blob/a14c678ec28ec804b37984e07a794b9c22acf737/src/vs/editor/standalone/browser/standaloneServices.ts#L326) 的 override

**Comment by bytemain** [comment]
是否可以采用一些微前端的方案？让 monaco 的 context 是我们可以控制的，这样我们就可以销毁它。
